From 7fe5ede85d467f4ecaa98ef76f789f781d4b99c2 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:19:17 -0400
Subject: [net] netfilter: nf_tables: skip transaction if no update flags in tables

Message-id: <20141021161917.22888.80920.stgit@localhost.localdomain>
Patchwork-id: 97975
O-Subject: [RHEL7 net PATCH 078/124] netfilter: nf_tables: skip transaction if no update flags in tables
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 63283dd21ed2bf25a71909a820ed3e8fe412e15d

 commit 63283dd21ed2bf25a71909a820ed3e8fe412e15d
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Fri Jun 27 18:51:39 2014 +0200

    netfilter: nf_tables: skip transaction if no update flags in tables

    Skip transaction handling for table updates with no changes in
    the flags. This fixes a crash when passing the table flag with all
    bits unset.

    Reported-by: Ana Rey <anarey@gmail.com>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index ab4566c..da5dc37 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -407,6 +407,9 @@ static int nf_tables_updtable(struct nft_ctx *ctx)
  if (flags & ~NFT_TABLE_F_DORMANT)
   return -EINVAL;
 
+ if (flags == ctx->table->flags)
+  return 0;
+
  trans = nft_trans_alloc(ctx, NFT_MSG_NEWTABLE,
     sizeof(struct nft_trans_table));
  if (trans == NULL)
-- 
1.7.1